src/ccl_cls.c

```diff
@@ -18,202 +18,271 @@ typedef struct{
-    // Return if computation failed
-    if (!cosmo->computed_power) return NAN;
-    psp_use=cosmo->data.p_nl;
+  if (psp == NULL) {
+    if (!cosmo->computed_nonlin_power) {
+      *status = CCL_ERROR_NONLIN_POWER_INIT;
+      ccl_cosmology_set_status_message(
+        cosmo,
+        "ccl_cls.c: ccl_angular_cl_limber(): non-linear power spctrum has not been computed!");
+      return;
+    }
+    psp_use = cosmo->data.p_nl;
   }
   else
-    psp_use=psp;
+    psp_use = psp;
 
-  int clastatus=0;
-  //Set up integrating function
-  integ_cl_par ipar;
-  gsl_function F;
-  ipar.l=l;
-  ipar.cosmo=cosmo;
-  ipar.trc1=trc1;
-  ipar.trc2=trc2;
-  ipar.psp=psp_use;
-  ipar.status=&clastatus;
-  F.function=&cl_integrand;
-  F.params=&ipar;
-
-  //Get integration limits
-  double lkmin,lkmax;
-  get_k_interval(cosmo,trc1,trc2,l,&lkmin,&lkmax);
-
-  //Integrate
-  double result,eresult;
-  int gslstatus;
-  gsl_integration_workspace *w=gsl_integration_workspace_alloc(cosmo->gsl_params.N_ITERATION);
-  gslstatus=gsl_integration_qag(&F, lkmin, lkmax, 0,
-                                cosmo->gsl_params.INTEGRATION_LIMBER_EPSREL,
-				cosmo->gsl_params.N_ITERATION,
-                                cosmo->gsl_params.INTEGRATION_LIMBER_GAUSS_KRONROD_POINTS,
-                                w, &result, &eresult);
-  gsl_integration_workspace_free(w);
-
-  // Test if a round-off error occured in the evaluation of the integral
-  // If so, try another integration function, more robust but potentially slower
-  if(gslstatus == GSL_EROUND) {
-    ccl_raise_gsl_warning(gslstatus, "ccl_cls.c: ccl_angular_cl_limber(): "
-			  "Default GSL integration failure, attempting backup method.");
-    gsl_integration_cquad_workspace *w_cquad= \
-      gsl_integration_cquad_workspace_alloc(cosmo->gsl_params.N_ITERATION);
-    size_t nevals=0;
-    gslstatus=gsl_integration_cquad(&F, lkmin, lkmax, 0,
-				    cosmo->gsl_params.INTEGRATION_LIMBER_EPSREL,
-				    w_cquad, &result, &eresult, &nevals);
+  #pragma omp parallel private(lind, clastatus, ipar, lkmin, lkmax, result, \
+                               eresult, gslstatus, w, w_cquad, nevals, l, F, \
+                               local_status) \
+                       shared(cosmo, trc1, trc2, l_out, cl_out, \
+                              nl_out, status, psp_use) \
+                       default(none)
+  {
+    w_cquad = NULL;
+    w = NULL;
+    local_status = *status;
+
+    if (local_status == 0) {
+      w = gsl_integration_workspace_alloc(cosmo->gsl_params.N_ITERATION);
+      if (w == NULL) {
+        local_status = CCL_ERROR_MEMORY;
+      }
+    }
+
+    if (local_status == 0) {
+      // Set up integrating function
+      ipar.cosmo = cosmo;
+      ipar.trc1 = trc1;
+      ipar.trc2 = trc2;
+      ipar.psp = psp_use;
+      ipar.status = &clastatus;
+      F.function = &cl_integrand;
+      F.params = &ipar;
+    }
+
+    #pragma omp for
```

Comment:
Have you tested if `schedule(dynamic)` helps here? E.g. high-ell and low-ell integrals take different amounts of time.